app/server/appsmith-plugins/postgresPlugin/src/main/java/com/external/plugins/PostgresPlugin.java

```diff
@@ -114,27 +115,41 @@ public PostgresPlugin(PluginWrapper wrapper) {
             
             return Mono.fromCallable(() -> {
 
-                Connection connectionFromPool;
-
-                try {
-                    connectionFromPool = getConnectionFromConnectionPool(connection, datasourceConfiguration);
-                } catch (StaleConnectionException e) {
-                    return Mono.error(e);
-                } catch (SQLException e) {
-                    return Mono.error(new AppsmithPluginException(AppsmithPluginError.PLUGIN_ERROR, e.getMessage()));
-                }
-
                 String query = actionConfiguration.getBody();
+                // Check for query parameter before performing the probably expensive fetch connection from the pool op.
                 if (query == null) {
                     return Mono.error(new AppsmithPluginException(AppsmithPluginError.PLUGIN_ERROR, "Missing required parameter: Query."));
                 }
+                
+                Connection connectionFromPool = null;
+
+                try {
+                    connectionFromPool = getConnectionFromConnectionPool(connection, datasourceConfiguration);
+                } catch (SQLException |StaleConnectionException e) {
+                    // The function can throw either StaleConnectionException or SQLException. The underlying hikari
+                    // library throws SQLException in case the pool is closed or there is an issue initializing
+                    // the connection pool which can also be translated in our world to StaleConnectionException
+                    // and should then trigger the destruction and recreation of the pool.
+                    return Mono.error(new StaleConnectionException());
```

Comment:
Can we return the actual exception when it is a `StaleConnectionException`? This way, the actual stack trace of the exception is not lost, just in case it's traceback is logged somewhere.

```suggestion
                    return Mono.error(e instanceof StaleConnectionException ? e : new StaleConnectionException());
```